### PR TITLE
Fix gradient accumulation and learning rate aggregation

### PR DESCRIPTION
**Pull request by jeffpicard**
This is PR'ed to the upstream [here](https://github.com/flairNLP/flair/pull/3583), more details there.

But requesting to merge to Ziprecruiter/main as well (i.e. to land sooner).

### CODE REVIEW

File: flair/trainers/plugins/functional/checkpoints.py [hunk 1]
```diff
@@ -30,8 +30,6 @@ def after_training_epoch(self, epoch, **kw):
             )
             model_name = "model_epoch_" + str(epoch) + ".pt"
             self.model.save(self.base_path / model_name, checkpoint=self.save_optimizer_state)
-            if torch.distributed.is_initialized():
-                torch.distributed.barrier()  # Prevent any process from loading a model until writing is complete
```
Comment (MattGPT-ai) [inline flair/trainers/plugins/functional/checkpoints.py:34]: Why are we removing this? I see another instance of having a barrier in _save_model, is this just redundant?

File: flair/trainers/trainer.py [hunk 2]
```diff
@@ -656,37 +656,43 @@ def train_custom(
...
-                            # identify dynamic embeddings (always deleted) on first sentence
-                            if dynamic_embeddings is None:
-                                dynamic_embeddings = identify_dynamic_embeddings(batch)
-
-                            # depending on memory mode, embeddings are moved to CPU, GPU or deleted
-                            store_embeddings(batch_step, embeddings_storage_mode, dynamic_embeddings)
+                        for batch_step_no, batch_step in enumerate(batch_steps):
+                            disable_gradient_sync = multi_gpu and batch_step_no < len(batch_steps) - 1
+                            grad_sync = self.ddp_model.no_sync() if disable_gradient_sync else contextlib.nullcontext()
+                            with grad_sync:
```
Comment (MattGPT-ai) [inline flair/trainers/trainer.py:662]: Just checking I'm getting what's going on here, this is where you are disabling gradient sync for all but the last batch step, and everything else is the same, just has an indent change, is that right?

**Review by MattGPT-ai — APPROVED** [review]
Looks awesome. Want to merge this and change our build to use this PR so I can test this in action

File: flair/trainers/trainer.py (hunk 2 above)
Comment (jeffpicard) [reply flair/trainers/trainer.py:662]: Exactly

File: flair/trainers/plugins/functional/checkpoints.py (hunk 1 above)
Comment (jeffpicard) [reply flair/trainers/plugins/functional/checkpoints.py:34]: It lead to a deadlock that froze the app (explanation in the upstream PR). It would be good muscle memory to have a barrier whenever you save, but isn't strictly needed since there isn't another process that's going to read the saved checkpoint file.
